src/object/obj_tx.c

```diff
@@ -640,154 +673,774 @@ dc_tx_open(tse_task_t *task)
 }
 
 struct tx_commit_cb_args {
-	struct dc_tx	*tcca_tx;
-	crt_rpc_t	*tcca_req;
+	struct dc_tx		*tcca_tx;
+	crt_rpc_t		*tcca_req;
+	daos_tx_commit_t	*tcca_args;
 };
 
 static int
 dc_tx_commit_cb(tse_task_t *task, void *data)
 {
-	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct tx_commit_cb_args *tcca = data;
 	struct dc_tx		 *tx = tcca->tcca_tx;
 	crt_rpc_t		 *req = tcca->tcca_req;
+	struct obj_cpd_out	 *oco = crt_reply_get(req);
+	uint64_t		 *sub_epochs;
 	int			  rc = task->dt_result;
+	bool			  locked = true;
 
 	D_MUTEX_LOCK(&tx->tx_lock);
 
+	if (rc == 0) {
+		rc = oco->oco_ret;
+		if (rc == 0) {
+			int	*sub_rets = oco->oco_sub_rets.ca_arrays;
+
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			rc = sub_rets[0];
+		}
+	}
+
 	if (rc == 0) {
 		tx->tx_status = TX_COMMITTED;
+		tx->tx_retry = 0;
+		dc_tx_cleanup(tx);
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		if (tx->tx_epoch.oe_value == 0)
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
+		if (tx->tx_epoch.oe_value == 0)
+			D_WARN("Server forgot to reply epoch for TX "DF_DTI"\n",
+			       DP_DTI(&tx->tx_id));
+
 		goto out;
 	}
 
 	tx->tx_status = TX_FAILED;
 
-	/* Required to restart the TX with newer epoch. */
-	if (rc == -DER_TX_RESTART)
+	/* Be required to restart the TX with newer epoch. */
+	if (rc == -DER_TX_RESTART) {
+		tx->tx_pm_ver = oco->oco_map_version;
+		task->dt_result = rc;
+		/* Set retry flag even if it is restarted. */
+		tx->tx_retry = 1;
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		/* FIXME: currently, we pack one DTX in one CPD RPC. */
+		if (tx->tx_epoch.oe_value < sub_epochs[0])
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
 		goto out;
+	}
+
+	D_ASSERTF(tx->tx_pm_ver >= oco->oco_map_version,
+		  "Invalid pm version for TX "DF_DTI": TX ver %u, RPC ver %u\n",
+		  DP_DTI(&tx->tx_id), tx->tx_pm_ver, oco->oco_map_version);
+
+	if (!obj_retry_error(rc)) {
+		task->dt_result = rc;
+		tx->tx_retry = 0;
+
+		goto out;
+	}
 
-	/* FIXME:
-	 * 1. Check pool map, if pool map is refreshed, restart TX.
-	 * 2. Retry the RPC if it is retriable error.
-	 * 3. Failout for other failures.
+	/* Keep the retry flag until TX is closed. */
+	tx->tx_retry = 1;
+	tcca->tcca_args->flags |= DTF_RETRY_COMMIT;
+
+	D_MUTEX_UNLOCK(&tx->tx_lock);
+	locked = false;
+
+	/* CPD RPC retry, do not need to refresh the pool map, because if
+	 * client pool map is stale, we should have hit '-DER_TX_RESTART'.
 	 */
```

Comment:
Unrelated to this client-side patch, how about using -DER_STALE on the wire for the stale pool map case? We could then convert -DER_STALEs to -DER_TX_RESTARTs if necessary on the client side. This would (a) be more consistent with existing stale pool map cases (I think?) and (b) provide clients with more info that might be useful in the future perhaps.

Reply:
Sometimes, server will not return -DER_STALE for stale pool map, instead, just returns the server known latest pool map version. It is the client duty to compare the returned pool map version and its local pool map version, and decide whether need to refresh its local pool map.

For distributed transaction case, if the client knows its local pool map is stale, it will refresh it, then ask the TX sponsor to restart the TX. If the server directly return -DER_TX_RESTART, then the client may not refresh its pool map.

BTW: because of some test failures, I will not refresh this patch right away to avoid wasting yours review time and test resource.